### PR TITLE
cli,pkg: use zap.Logger

### LINKED ISSUE #284: Change logging to use uber zap
As proposed in [#247](https://github.com/nspcc-dev/neo-go/issues/247) we have to change fmt.println logging to log but at the same time suggest to use [uber/zap](https://github.com/uber-go/zap) since it has better performance compare to logrus which is used in master

### PR DESCRIPTION
**Pull request by fyrchik**
Closes #284 .

### CODE REVIEW

File: pkg/core/blockchain.go [hunk 1]
```diff
@@ -80,13 +80,21 @@ type Blockchain struct {
...
 
 // NewBlockchain returns a new blockchain object the will use the
 // given Store as its underlying storage.
-func NewBlockchain(s storage.Store, cfg config.ProtocolConfiguration) (*Blockchain, error) {
+func NewBlockchain(s storage.Store, cfg config.ProtocolConfiguration, log *zap.Logger) (*Blockchain, error) {
+	if log == nil {
+		log = zap.NewNop()
+	}
+
+	icLog = log
```
Comment (roman-khimov) [inline pkg/core/blockchain.go]: What's this?

File: pkg/core/blockchain.go [hunk 2]
```diff
@@ -131,7 +140,7 @@ func (bc *Blockchain) init() error {
 	// At this point there was no version found in the storage which
 	// implies a creating fresh storage with the version specified
 	// and the genesis block as first block.
-	log.Infof("restoring blockchain with version: %s", version)
+	bc.log.Info("restoring blockchain with version", zap.String("version", version))
```
Comment (roman-khimov) [inline pkg/core/blockchain.go]: I think `restoring blockchain` will be enough here, given that we log `version` separately now.

File: pkg/core/interop_system.go [hunk 3]
```diff
@@ -5,6 +5,8 @@ import (
 	"fmt"
 	"math"
 
+	"go.uber.org/zap"
+
```
Comment (roman-khimov) [inline pkg/core/interop_system.go]: Useless blank line.

File: pkg/core/blockchain.go (hunk 1 above)
Comment (roman-khimov) [reply pkg/core/blockchain.go]: Now I do see that it's defined for interops, but I don't think it's a good idea, better pass it into the `newInteropContext()`.

Comment (roman-khimov) [reply pkg/core/blockchain.go]: Or take it from bc.log.

File: pkg/core/storage/boltdb_store.go [hunk 4]
```diff
@@ -100,17 +99,15 @@ func (s *BoltDBStore) PutBatch(batch Batch) error {
 
 // Seek implements the Store interface.
 func (s *BoltDBStore) Seek(key []byte, f func(k, v []byte)) {
-	err := s.db.View(func(tx *bbolt.Tx) error {
+	// no error will be returns as inner function returns no error
```
Comment (roman-khimov) [inline pkg/core/storage/boltdb_store.go]: I wouldn't be so sure about that, you don't know what kind of error `View` itself can return. Just look at the implementation: https://github.com/etcd-io/bbolt/blob/master/db.go#L708

File: pkg/network/payload/headers.go [hunk 5]
```diff
@@ -22,7 +21,6 @@ func (p *Headers) DecodeBinary(br *io.BinReader) {
 
 	// C# node does it silently
 	if lenHeaders > MaxHeadersAllowed {
-		log.Warnf("received %d headers, capping to %d", lenHeaders, MaxHeadersAllowed)
```
Comment (roman-khimov) [inline pkg/network/payload/headers.go]: Not sure it's a good idea to drop it. Downplaying it from warning may be useful, but dropping completely?..

File: pkg/network/tcp_transport.go [hunk 6]
```diff
@@ -80,10 +82,9 @@ func (t *TCPTransport) handleConn(conn net.Conn) {
 
 	// When a new peer is connected we send out our version immediately.
 	if err := t.server.sendVersion(p); err != nil {
-		log.WithFields(log.Fields{
-			"addr": p.RemoteAddr(),
-		}).Error(err)
+		t.log.Error("error on sendVersion", zap.Stringer("addr", p.RemoteAddr()), zap.Error(err))
 	}
+
```
Comment (roman-khimov) [inline pkg/network/tcp_transport.go]: Useless blank line.

File: cli/server/server.go [hunk 7]
```diff
@@ -119,26 +120,35 @@ func getConfigFromContext(ctx *cli.Context) (config.Config, error) {
...
 // If logPath is configured -- function creates dir and file for logging.
-func handleLoggingParams(ctx *cli.Context, cfg config.ApplicationConfiguration) error {
+func handleLoggingParams(ctx *cli.Context, cfg config.ApplicationConfiguration) (*zap.Logger, error) {
+	level := zapcore.InfoLevel
 	if ctx.Bool("debug") {
-		log.SetLevel(log.DebugLevel)
+		level = zapcore.DebugLevel
 	}
 
+	cc := zap.NewDevelopmentConfig()
```
Comment (im-kulikov) [inline cli/server/server.go]: Why you’re used development logger?

File: pkg/rpc/request.go [hunk 8]
```diff
@@ -85,32 +85,34 @@ func (r Request) WriteErrorResponse(w http.ResponseWriter, err error) {
...
 	}
 
-	r.writeServerResponse(w, response)
+	s.writeServerResponse(r, w, response)
 }
 
-func (r Request) writeServerResponse(w http.ResponseWriter, response Response) {
+func (s *Server) writeServerResponse(r *Request, w http.ResponseWriter, response Response) {
 	w.Header().Set("Content-Type", "application/json; charset=utf-8")
 	if r.enableCORSWorkaround {
```
Comment (roman-khimov) [inline pkg/rpc/request.go]: After this changes you should remove `enableCORSWorkaround` from the `Request`.

File: pkg/rpc/server.go [hunk 9]
```diff
@@ -7,6 +7,8 @@ import (
 	"net/http"
 	"strconv"
 
+	"go.uber.org/zap"
+
```
Comment (roman-khimov) [inline pkg/rpc/server.go]: Useless blank line.

File: pkg/core/interop_system.go (hunk 3 above)
Comment (im-kulikov) [inline pkg/core/interop_system.go]: Empty line

File: pkg/network/helper_test.go [hunk 10]
```diff
@@ -7,6 +7,8 @@ import (
 	"testing"
 	"time"
 
+	"go.uber.org/zap"
+
```
Comment (im-kulikov) [inline pkg/network/helper_test.go]: Empty line

File: cli/server/server.go [hunk 11]
```diff
@@ -302,12 +315,12 @@ func startServer(ctx *cli.Context) error {
...
 	serverConfig := network.NewServerConfig(cfg)
 
-	chain, prometheus, pprof, err := initBCWithMetrics(cfg)
+	chain, prometheus, pprof, err := initBCWithMetrics(cfg, log)
 	if err != nil {
 		return err
 	}
 
-	server := network.NewServer(serverConfig, chain)
+	server := network.NewServer(serverConfig, chain, nil)
```
Comment (roman-khimov) [inline cli/server/server.go]: `nil` logger for a server? I don't think it's a good idea. We should have the same logging level after this changeset as we had before (not counting consensus debugging, of course, that should go away).

File: pkg/network/metrics/metrics.go [hunk 12]
```diff
@@ -24,27 +25,30 @@ type Config struct {
...
+	ms.log.Info("shutting down service", zap.String("endpoint", ms.Addr))
 	err := ms.Shutdown(context.Background())
 	if err != nil {
-		log.Fatalf("can't shut down %s service", ms.serviceType)
+		ms.log.Panic("can't shut down service")
 	}
 }
+
+func getLogger(name string) *zap.Logger {
+	log, err := zap.NewDevelopment()
```
Comment (im-kulikov) [inline pkg/network/metrics/metrics.go]: Why you’re used development logger?

File: cli/server/server.go (hunk 11 above)
Comment (roman-khimov) [reply cli/server/server.go]: Also note that we should respect users debug choice for all packages at the moment (I guess we'll need to implement per-package logging choices in the future, but not now).

File: pkg/rpc/request.go [hunk 13]
```diff
@@ -85,32 +85,34 @@ func (r Request) WriteErrorResponse(w http.ResponseWriter, err error) {
...
-		"method": r.Method,
+	logFields := []zap.Field{
+		zap.Error(jsonErr.Cause),
+		zap.String("method", r.Method),
 	}
+
 	params, err := r.Params()
 	if err == nil {
-		logFields["params"] = *params
+		logFields = append(logFields, zap.Any("params", params))
```
Comment (im-kulikov) [inline pkg/rpc/request.go]: Don’t use any, it produced huge allocs

**Comment by im-kulikov** [comment]
NewNop is bad idea. Just return error if logger wasn’t received on yours side

File: cli/server/server.go (hunk 11 above)
Comment (fyrchik) [reply cli/server/server.go]: oops, it is a mistake probably, `nil` was intended only for tests.

File: pkg/rpc/request.go (hunk 13 above)
Comment (fyrchik) [reply pkg/rpc/request.go]: What do you suggest to use instead?

File: pkg/network/metrics/metrics.go (hunk 12 above)
Comment (fyrchik) [reply pkg/network/metrics/metrics.go]: I am not a logger >_<

Comment (fyrchik) [reply pkg/network/metrics/metrics.go]: On a serious note: it provides the output which is more human readable, production config just ouputs json. I will deal with this in future.

File: pkg/network/payload/headers.go (hunk 5 above)
Comment (fyrchik) [reply pkg/network/payload/headers.go]: I think it is better when encode/decode code is free of side effects.

I have come up with 2 other solutions:
1. Make a field `bool Headers.capped` and do some logging in caller.
2. Set an error in `BinaryReader`, then log and nullify it if it is `ErrMaxHeadersExceeded`

Comment (roman-khimov) [reply pkg/network/payload/headers.go]: Technically it'd be correct to set a `br.Err` because we've not decoded the payload completely here, but I'm not sure how it would complicate the error path above this decoder.

Comment (fyrchik) [reply pkg/network/payload/headers.go]: I set `Received` field on decoding and later check if it is consistent with provided array. This seems the cleanest solution to me, as it doesn't pollute main code path with the details of the specific command.

File: pkg/rpc/server.go [hunk 14]
```diff
@@ -44,38 +45,36 @@ func NewServer(chain core.Blockchainer, conf config.RPCConfig, coreServer *netwo
 		chain:      chain,
 		config:     conf,
 		coreServer: coreServer,
+		log:        zap.NewNop(),
```
Comment (roman-khimov) [inline pkg/rpc/server.go]: Whoa. It's not a testing code!

File: pkg/core/interops.go [hunk 15]
```diff
@@ -10,6 +10,8 @@ package core
 import (
 	"sort"
 
+	"go.uber.org/zap"
+
```
Comment (roman-khimov) [inline pkg/core/interops.go]: Useless. Blank. Line.

File: pkg/network/server.go [hunk 16]
```diff
@@ -80,21 +82,27 @@ func randomID() uint32 {
 }
 
 // NewServer returns a new Server, initialized with the given configuration.
-func NewServer(config ServerConfig, chain core.Blockchainer) *Server {
+func NewServer(config ServerConfig, chain core.Blockchainer, log *zap.Logger) *Server {
+	if log == nil {
+		log = zap.NewNop()
```
Comment (roman-khimov) [inline pkg/network/server.go]: I think we're better being consistent about all of the `new*` functions, at the moment `NewBlockchain` fails with `nil` logger and all the others succeed. IMO, it's OK to succeed, but this behavior should be the same for all functions like this.

File: pkg/network/metrics/pprof.go [hunk 17]
```diff
@@ -18,9 +18,12 @@ func NewPprofService(cfg Config) *Service {
...
 	return &Service{
-		&http.Server{
+		Server: &http.Server{
 			Addr:    cfg.Address + ":" + cfg.Port,
 			Handler: handler,
-		}, cfg, "Pprof",
+		},
+		config:      cfg,
+		serviceType: "Pprof",
+		log:         getLogger("Pprof"),
```
Comment (roman-khimov) [inline pkg/network/metrics/pprof.go]: Why can't get a logger via parameters?

Comment (fyrchik) [reply pkg/network/metrics/pprof.go]: We can!

File: pkg/network/metrics/metrics.go (hunk 12 above)
Comment (fyrchik) [reply pkg/network/metrics/metrics.go]: Is there any reason for using Production config now? I'd prefer to see time and durations human-readable format.

File: pkg/rpc/server.go (hunk 14 above)
Comment (roman-khimov) [reply pkg/rpc/server.go]: So, is there a reason for it to be nop?

File: pkg/core/interops.go (hunk 15 above)
Comment (roman-khimov) [reply pkg/core/interops.go]: Still there.

File: pkg/network/metrics/pprof.go (hunk 17 above)
Comment (roman-khimov) [reply pkg/network/metrics/pprof.go]: Well, we should then do that for both services, let them accept logger via `NewXXX()` parameters.

File: pkg/network/metrics/metrics.go (hunk 12 above)
Comment (im-kulikov) [reply pkg/network/metrics/metrics.go]: Just use “console” format

File: pkg/rpc/request.go (hunk 13 above)
Comment (im-kulikov) [reply pkg/rpc/request.go]: You have more than one option to use.. int/string/other

File: pkg/consensus/consensus_test.go [hunk 18]
```diff
@@ -3,6 +3,8 @@ package consensus
 import (
 	"testing"
 
+	"go.uber.org/zap"
+
```
Comment (volekerb) [inline pkg/consensus/consensus_test.go]: space could be removed

File: pkg/core/interop_neo_test.go [hunk 19]
```diff
@@ -4,6 +4,8 @@ import (
 	"math/big"
 	"testing"
 
+	"go.uber.org/zap"
+
```
Comment (volekerb) [inline pkg/core/interop_neo_test.go]: empty space

File: pkg/core/storage/boltdb_store.go [hunk 20]
```diff
@@ -108,8 +107,9 @@ func (s *BoltDBStore) Seek(key []byte, f func(k, v []byte)) {
 		}
 		return nil
 	})
+
 	if err != nil {
-		log.Error("error while executing seek in boltDB")
+		panic(err)
```
Comment (volekerb) [inline pkg/core/storage/boltdb_store.go]: hm, now it's panic.. I guess this one should be in separate commit since it changes logic

File: cli/server/server.go (hunk 7 above)
Comment (roman-khimov) [reply cli/server/server.go]: I think we're better using production config by default. If you need something for development --- either keep a local patch for it or make it configurable (effectively moving this patch from code patch to config patch).

File: pkg/network/metrics/prometheus.go [hunk 21]
```diff
@@ -3,18 +3,27 @@ package metrics
 import (
 	"net/http"
 
+	"go.uber.org/zap"
+
```
Comment (roman-khimov) [inline pkg/network/metrics/prometheus.go]: Useless blank line.

File: pkg/core/storage/boltdb_store.go (hunk 20 above)
Comment (roman-khimov) [reply pkg/core/storage/boltdb_store.go]: +1.

File: pkg/network/payload/headers.go [hunk 22]
```diff
@@ -19,6 +21,8 @@ const (
 func (p *Headers) DecodeBinary(br *io.BinReader) {
 	lenHeaders := br.ReadVarUint()
 
+	p.Received = lenHeaders
```
Comment (roman-khimov) [inline pkg/network/payload/headers.go]: What if we're to use `br.Err` to pass it as an error? I think it would still be possible to recover in `handleConn()`. I just don't like introducing artificial fields in a structure that is supposed to represent a network message payload.

Comment (fyrchik) [reply pkg/network/payload/headers.go]: If we will do it via `br.Err` then we should expect expect error everytime we use `DecodeBinary` on a message containing `Headers`. I thought this could lead to some easy-to-miss bugs.
Maybe this is not a problem, though as `Headers` is used only as a last field of a single message.

File: pkg/rpc/request.go (hunk 13 above)
Comment (fyrchik) [reply pkg/rpc/request.go]: What do you suggest to use here?

Comment (im-kulikov) [reply pkg/rpc/request.go]: `for range` throw Params() and switch case by type?

File: cli/server/server.go (hunk 7 above)
Comment (im-kulikov) [reply cli/server/server.go]: fix that

File: pkg/network/payload/headers.go (hunk 22 above)
Comment (roman-khimov) [reply pkg/network/payload/headers.go]: Well, we only have one point where we decode this message, if we're to have more then maybe they'll need to have different behaviour wrt this error. Even if not, passing it as an error makes more sense to me anyway.

File: pkg/rpc/request.go (hunk 13 above)
Comment (fyrchik) [reply pkg/rpc/request.go]: Does it make more allocs than `logrus` in similar case?

Comment (im-kulikov) [reply pkg/rpc/request.go]: nop, but... It encodes message as JSON and then writes to stdout, it can produce huge memallocs on an empty sit

**Comment by im-kulikov** [comment]
can you replace all `zap.NewNop` with `https://godoc.org/go.uber.org/zap/zaptest`

File: pkg/rpc/request.go (hunk 13 above)
Comment (fyrchik) [reply pkg/rpc/request.go]: I don't see a necessity to do it right now. We will do it in another PR.

Comment (roman-khimov) [reply pkg/rpc/request.go]: OK, it's a microoptimization, we may not bother for now if it simplifies transition from logrus.

File: cli/server/server.go [hunk 23]
```diff
@@ -119,32 +120,44 @@ func getConfigFromContext(ctx *cli.Context) (config.Config, error) {
...
+	cc.Encoding = "console"
+	cc.Level = zap.NewAtomicLevelAt(level)
+
 	if logPath := cfg.LogPath; logPath != "" {
 		if err := io.MakeDirForFile(logPath, "logger"); err != nil {
-			return err
+			return nil, err
 		}
-		f, err := os.Create(logPath)
+		_, err := os.Create(logPath)
```
Comment (roman-khimov) [inline cli/server/server.go]: Aren't you leaking a file descriptor here?

**Comment by fyrchik** [comment]
> [ae27d13](https://github.com/nspcc-dev/neo-go/commit/ae27d138625e1c0e85be9fd1dea00eca11ae008e) has a wrong commit message now and we have tests failing for some reason.

They all pass locally, I will dig into this.
